packages/core/src/types/languages.ts

```diff
@@ -1,47 +1,44 @@
+import type { CachedFactory } from "cached-factory";
+
+import type { AnyOptionalSchema } from "./shapes.js";
+
 import { CommentDirective } from "./directives.js";
-import { PromiseOrSync } from "./promises.js";
-import { FileReport, NormalizedReport } from "./reports.js";
 import {
-	AnyRule,
-	AnyRuleDefinition,
-	Rule,
-	RuleAbout,
-	RuleDefinition,
-} from "./rules.js";
-import { AnyOptionalSchema, InferredObject } from "./shapes.js";
+	FileReport,
+	NormalizedReport,
+	type ReportMessageData,
+} from "./reports.js";
+import { Rule, RuleAbout, RuleDefinition, type RuleRuntime } from "./rules.js";
 
 export type AnyLanguage = Language<object, object>;
 
-export interface CreateRule<AstNodesByName, ContextServices extends object> {
-	<const About extends RuleAbout, const MessageId extends string>(
-		definition: RuleDefinition<
-			About,
-			AstNodesByName,
-			ContextServices,
-			MessageId,
-			undefined
-		>,
-	): Rule<About, AstNodesByName, ContextServices, MessageId, undefined>;
-
-	<
-		const About extends RuleAbout,
-		const MessageId extends string,
-		const OptionsSchema extends AnyOptionalSchema,
-	>(
-		definition: RuleDefinition<
-			About,
-			AstNodesByName,
-			ContextServices,
-			MessageId,
-			OptionsSchema
-		>,
-	): Rule<About, AstNodesByName, ContextServices, MessageId, OptionsSchema>;
-}
+export type CreateRule<AstNodesByName, ContextServices extends object> = <
+	const About extends RuleAbout,
+	const MessageId extends string,
+	const FileContext extends object | undefined = undefined,
+	const OptionsSchema extends AnyOptionalSchema | undefined = undefined,
+>(
+	definition: RuleDefinition<
+		About,
+		AstNodesByName,
+		ContextServices,
+		FileContext,
+		MessageId,
+		OptionsSchema
+	>,
+) => Rule<
+	About,
+	AstNodesByName,
+	ContextServices,
+	FileContext,
+	MessageId,
+	OptionsSchema
+>;
 
-export interface Language<AstNodesByName, ContextServices extends object>
-	extends LanguageDefinition {
```

Comment:
Is this important to maintain? It prevents having a different type for `prepareFromFile` in `LanguageFileFactory` `LanguageFileFactoryDefinition`